src/linker/Linker/Annotations.cs

```diff
@@ -553,29 +552,45 @@ public bool SetPreservedStaticCtor (TypeDefinition type)
 			return marked_types_with_cctor.Add (type);
 		}
 
-		public bool HasRequiresUnreferencedCodeOnTypeHierarchy (TypeDefinition type)
+		/// <summary>
+		/// Looks if there is a RequiresUnreferencedCodeAttribute on the <paramref name="type"/> or any of its declaring types
+		/// </summary>
+		/// <param name="type">Type to start the search of the attribute</param>
+		/// <returns>True if it was able to find the attribute on the type or any of its declaring types, false otherwise</returns>
+		public bool HasEffectiveRequiresUnreferencedCodeOnType (TypeDefinition type)
 		{
-			if (!ruc_on_type_hierarchy.TryGetValue (type, out var hasRucOnTypeHierarchy))
-				return TryGetRequiresUnreferencedCodeAttributeOnTypeHierarchy (type, out var _);
-			return hasRucOnTypeHierarchy is not null;
+			if (HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (type))
+				return true;
+			while (type.DeclaringType != null) {
+				if (HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (type))
+					return true;
+				type = type.DeclaringType;
+			}
```

Comment:
For simplicity I would just call TryGet... here and discard the attribute value.